File: data_structures/trie/trie_bfs.go
package trie

type Trie struct {
	IsWord bool
	Next   map[rune]*Trie
}

func NewTrie() *Trie {
	return &Trie{Next: make(map[rune]*Trie)}
}

func (t *Trie) Insert(word string) {
	currTrie := t

	for _, char := range word {
		if _, ok := currTrie.Next[char]; !ok {
			currTrie.Next[char] = NewTrie()
		}
		currTrie = currTrie.Next[char]
	}

	currTrie.IsWord = true
}

func (t *Trie) Search(word string) bool {
	currTrie := t

	for _, char := range word {

		if _, ok := currTrie.Next[char]; !ok {
			return false
		}

		currTrie = currTrie.Next[char]
	}

	return currTrie.IsWord
}

func (t *Trie) FindPrefix(prefix string) bool {
	currTrie := t

	for _, char := range prefix {

		if _, ok := currTrie.Next[char]; !ok {
			return false
		}

		currTrie = currTrie.Next[char]
	}

	return true
}

File: data_structures/trie/trie_bfs.cpp
#include <iostream>
#include <unordered_map>

class Trie {
public:
  Trie() {
    m_children.reserve(26);
  }
  ~Trie () {
    for (auto& pair : m_children) {
      delete pair.second;
    }
  }

  void insert(std::string word) {
    Trie* currNode = this;
    
    for (const char& c : word ) {  
      if (currNode->m_children.find(c) == currNode->m_children.end()) {
        currNode->m_children[c] = new Trie();
      }
      currNode = currNode->m_children[c];
    }

    currNode->m_isWord = true;
  }

  bool search(std::string word) {
    Trie* currNode = this;
    
    for (const char& c : word ) {  
      if (currNode->m_children.find(c) == currNode->m_children.end()) {
        return false;
      }
      currNode = currNode->m_children[c];
    }

    return currNode->m_isWord;
  }

  bool findPrefix(std::string word) {
    Trie* currNode = this;
    
    for (const char& c : word ) {  
      if (currNode->m_children.find(c) == currNode->m_children.end()) {
        return false;
      }
      currNode = currNode->m_children[c];
    }

    return true;
  }

private:
  bool m_isWord = false;
  std::unordered_map<char, Trie*> m_children;
};